src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -0,0 +1,703 @@
+                        $"Activity {act.Context.Info.ActivityType} is not registered on this worker," +
+                        $" available activities: {availStr}",
+                        errorType: "NotFoundError");
+                }
+
+                // Deserialize arguments. If the input is less than the required parameter count, we
+                // error.
+                var paramInfos = defn.Delegate.Method.GetParameters();
+                if (tsk.Start.Input.Count < paramInfos.Length &&
+                    !paramInfos[tsk.Start.Input.Count].HasDefaultValue)
+                {
+                    throw new ApplicationFailureException(
+                        $"Activity {act.Context.Info.ActivityType} given {tsk.Start.Input.Count} parameter(s)," +
+                        " but more than that are required by the signature");
+                }
+                // Zip the params and input and then decode each. It is intentional that we discard
+                // extra input arguments that the signature doesn't accept.
+                var paramVals = new List<object?>(tsk.Start.Input.Count);
+                try
+                {
+                    foreach (var (input, paramInfo) in tsk.Start.Input.Zip(paramInfos, (a, b) => (a, b)))
+                    {
+                        var paramVal = await worker.Client.Options.DataConverter.ToValueAsync(
+                            input, paramInfo.ParameterType).ConfigureAwait(false);
+                        paramVals.Add(paramVal);
+                    }
+                }
+                catch (Exception e)
+                {
+                    throw new ApplicationFailureException(
+                        "Failed decoding parameters", e);
+                }
+                // Append default parameters if needed
+                for (var i = tsk.Start.Input.Count; i < paramInfos.Length; i++)
+                {
+                    paramVals.Add(paramInfos[i].DefaultValue);
+                }
+
+                // Build the interceptor impls, chaining each interceptor in reverse
+                var inbound = worker.Interceptors.Reverse().Aggregate(
+                    (ActivityInboundInterceptor)new InboundImpl(),
+                    (v, impl) => impl.InterceptActivity(v));
+                // Initialize with outbound
+                inbound.Init(new OutboundImpl(this));
+
+                // Execute and put result on completed
+                var result = await inbound.ExecuteActivityAsync(new(
+                    Delegate: defn.Delegate,
+                    Parameters: paramVals.ToArray(),
+                    Headers: tsk.Start.HeaderFields)).ConfigureAwait(false);
+
+                completion.Result.Completed = new();
+                // As a special case, ValueTuple is considered "void"
+                if (result is not ValueTuple)
+                {
+                    completion.Result.Completed.Result =
```

Comment:
AFAIR, we decided that "null" result should end up as a null payload even though the behavior is inconsistent in the other SDKs.

Reply:
A null result will end up as a null payload, but this is a "void" which while it can't be differentiated in TS and Python from null/undefined/None, it can in Java and .NET.

Reply:
I'm not sure Java makes that distinction but I'm fine with this behavior for .NET.